src/toil/cwl/cwltoil.py

```diff
@@ -446,92 +450,110 @@ class ToilPathMapper(PathMapper):
-                    resolved = schema_salad.ref_resolver.uri_file_path(
-                        resolved)
-                self._pathmap[loc] = MapperEnt(
+                    resolved = schema_salad.ref_resolver.uri_file_path(resolved)
+                self._pathmap[path] = MapperEnt(
                     resolved, tgt, "WritableFile" if copy else "File", staged)
                 self.visitlisting(obj.get("secondaryFiles", []),
                                   stagedir, basedir, copy=copy, staged=staged)
 
 
 class ToilCommandLineTool(cwltool.command_line_tool.CommandLineTool):
-    """Subclass the cwltool command line tool to provide the custom
-    Toil.PathMapper.
+    """Subclass the cwltool command line tool to provide the custom Toil.PathMapper."""
+    def make_path_mapper(self,
+                         reffiles: List[Any],
+                         stagedir: str,
+                         runtimeContext: cwltool.context.RuntimeContext,
+                         separateDirs: bool) -> cwltool.pathmapper.PathMapper:
+        return ToilPathMapper(reffiles,
+                              runtimeContext.basedir,
+                              stagedir,
+                              separateDirs,
+                              runtimeContext.toil_get_file)
+
+
+def toil_make_tool(toolpath_object: MutableMapping[str, Any],
+                   loading_context: cwltool.context.LoadingContext) -> cwltool.command_line_tool.CommandLineTool:
     """
-
-    def make_path_mapper(self, reffiles, stagedir, runtimeContext,
-                         separateDirs):
-        return ToilPathMapper(
-            reffiles, runtimeContext.basedir, stagedir, separateDirs,
-            runtimeContext.toil_get_file)
-
-
-def toil_make_tool(toolpath_object, loading_context):
-    """Factory function passed to load_tool() which creates instances of the
+    Factory function passed to load_tool() which creates instances of the
     custom ToilCommandLineTool.
-
     """
-
-    if isinstance(toolpath_object, Mapping) \
-            and toolpath_object.get("class") == "CommandLineTool":
+    if isinstance(toolpath_object, Mapping) and toolpath_object.get("class") == "CommandLineTool":
         return ToilCommandLineTool(toolpath_object, loading_context)
     return cwltool.workflow.default_make_tool(toolpath_object, loading_context)
 
 
 class ToilFsAccess(cwltool.stdfsaccess.StdFsAccess):
-    """Custom filesystem access class which handles toil filestore references.
-
-    """
-
-    def __init__(self, basedir, file_store=None):
+    """Custom filesystem access class which handles toil filestore references."""
+    def __init__(self, basedir: str, file_store: AbstractFileStore = None):
         self.file_store = file_store
         super(ToilFsAccess, self).__init__(basedir)
 
-    def _abs(self, path):
+    def exists(self, path: str) -> bool:
+        # toil's _abs() throws errors when files are not found and cwltool's _abs() does not
```

Comment:
Why not fix this in the `_abs` method itself?

Reply:
@mr-c I originally only modified `_abs`.  The problem is that it translates from `toilfs:` to `file://`, so if it doesn't exist, it can only return a randomly generated `file://` path.  @adamnovak had a concern that a call to `_abs` to predict where a file currently isn't but _will_ exist couldn't work then (unless we started creating deterministic paths with `uuid5` or something).